feat: protected routes defined from nuxt config

DanielRivers opened the pull request:
This adds the ability to protect routes based on permissions of the logged in user

this is all handled by the `auth-logged-in` middleware

example configuration

```
    routeRules: [
      {
        "/dashboard": {
		  appMiddleware: ['auth-logged-in'],  // 3.11 or greater
          kinde: {
            permissions: ['example_permission'],
            redirectUrl: "/"
          }
        },
       "/admin/**": {
		  appMiddleware: ['auth-logged-in'], // 3.11 or greater
          kinde: {
            permissions: ['admin_user'],
            redirectUrl: "/"
          }
        }
      }
    ]
```

This will protect the route `protected` to only users with the permission of `example_permission` and all routes under `admin` need to have `admin_user` permission

An array of permissions can be defined, along with an optional `redirectUrl`.  If not redirectUrl is supplied the user will be presented with a `401` error as per the current behaviour.

For versions 3.10 and lower you have to define the middleware on the you want to be protected 

```
definePageMeta({
  middleware: ['auth-logged-in'],
})
```


This is a built in solution for https://github.com/nuxt-modules/kinde/issues/45

danielroe commented:
What about using custom `routeRules` for this?

DanielRivers commented:
@danielroe I can't see what I have done to break the virtual file `src/runtime/server/utils/client.ts`

None of my seem to marry up to cause the error.

danielroe commented on src/kit.ts:
```diff
@@ -0,0 +1,23 @@
+import { addTemplate, createResolver, useNuxt } from "@nuxt/kit"
+import { relative } from 'pathe'
+
+export function extendTypes(module: string, template: (options: { typesPath: string }) => string | Promise<string>) {
```
why are we adding this utility? was there an issue with the `addTypeTemplate` utility from nuxt/kit?

DanielRivers replied:
I wasn't aware of this, I can switch out.

I used ogImage as a reference for adding the route rules, it has the same method, pleased there is a method in nuxt/kit 

https://github.com/nuxt-modules/og-image/blob/e139fdae6e62a1f008c2086b4ce8dc3946a23b48/src/kit.ts#L36C2-L39C3

DanielRivers replied:
Switched out

danielroe commented:
One question is whether we instead move forward with https://github.com/nuxt/nuxt/pull/25841 rather than implementing this specifically in this module. That would allow kinde middleware to be applied in a more generic way. Do you see any limitations in that PR?

DanielRivers commented:
> One question is whether we instead move forward with [nuxt/nuxt#25841](https://github.com/nuxt/nuxt/pull/25841) rather than implementing this specifically in this module. That would allow kinde middleware to be applied in a more generic way. Do you see any limitations in that PR?

@danielroe That looks good, I was hoping to find a wa to trigger the middleware without having to add `middleware: ['auth-logged-in']` to every page. 

one would assume that the rest of the route payload would be accessible so can continue to use the following style of config and extend as needed?

```
    "/protected": {
      kinde: {
        permissions: ['example_permission'],
        redirectUrl: '/',
      }
    },
```

I can pull the PR and have a look in more detail, but if I understand right not a huge amount would change just really how its triggered?

DanielRivers commented:
@danielroe What changes are needed to get this over the line?

DanielRivers commented:
@danielroe Mostly your changes looked good, I had to change slightly the implementation of `getRouteRules` as it crashed when accessing any protected route.

I have tested on both direct access and using NuxtLink.

DanielRivers commented:
@danielroe What can we do to get this PR through?  We are seeing an increased usage of the module and route protection is a common question.

danielroe commented:
@DanielRivers As I mentioned previously, I don't believe this PR currently would work work on client-side navigation. For example, this code:

```ts
const usersPermissions = await nuxt.ssrContext!.event.context.kinde.getPermissions()
```

will not work in the browser.

DanielRivers commented:
@danielroe I have implemented client support, added an access endpoint which checks if the logged in user has access to the route, returns the returnUrl and access boolean.

danielroe commented on src/module.ts:
```diff
@@ -146,6 +147,13 @@ export default defineNuxtModule<ModuleOptions>({
         || resolver.resolve('./runtime/server/api/logout.get'),
     })
 
+    addServerHandler({
+      route: '/api/access',
+      handler:
+        options.handlers?.access
+        || resolver.resolve('./runtime/server/api/access.post'),
+    })
```
I think we could avoid adding the handler if there are no route rules that would take advantage of it.

DanielRivers commented:
@danielroe little nudge on a review 🙏

DanielRivers commented:
@danielroe Can this get a review please?

danielroe commented on src/runtime/middleware/auth-logged-in.ts:
```diff
@@ -29,6 +33,16 @@ export default defineNuxtRouteMiddleware(async () => {
+    }
+    return
+  }
+
```
I feel this should be done entirely on the client side rather than requiring an extra request every time someone tries to navigate to a given route.

are the permissions exposed/accessible to the client?

DanielRivers replied:
They're currently not as they are only on the token, I can certainly fetch and store in memory store on client when attempting to use and not there yet.

danielroe commented on src/module.ts:
```diff
@@ -59,6 +61,7 @@ export default defineNuxtModule<ModuleOptions>({
       register: '/api/register',
       health: '/api/health',
       logout: '/api/logout',
+      access: '/api/access',
```
FWIW, I think it would make sense to use a `/_kinde` prefix or something for e.g. `/_kinde/access` ... as these are quite generic and could easily overlap with user endpoints.

DanielRivers replied:
@danielroe These settings are default and are overridable in the config, reluctant to change at this point as it would be a breaking change when its not required. I would note that I agree with you, should possibly review later, but not had anyone raise this as an issue yet.

danielroe commented on src/runtime/middleware/auth-logged-in.ts:
```diff
@@ -1,13 +1,53 @@
+    return
+  }
+
+  if (!nuxt.$auth.loggedIn) {
+    return denyAccess(kindeConfig?.redirectUrl)
+  }
```
can users access pages while being logged out? if not, this should be moved earlier to L18 so we don't need to perform a network request or get the route rules

danielroe commented on src/runtime/middleware/auth-logged-in.ts:
```diff
@@ -1,13 +1,53 @@
+    return
+  }
+  const nuxt = useNuxtApp()
+  const { kinde: kindeConfig } = await getRouteRules({ path: useRoute().path }) as NitroRouteRules
```
user-defined route rules are only available on the server - so this will not provide any protection on the client

danielroe commented on src/runtime/server/api/access.post.ts:
```diff
@@ -0,0 +1,38 @@
+  const { kinde: kindeSettings, ...rest } = useRuntimeConfig()
+  const body = await readBody(event)
+
+  const routeRules: NitroRouteRules['kinde'] = rest.nitro?.routeRules?.[body.path]?.kinde
```
we want to use a route rules matcher here as they are nested, so a parent matcher can set permissions on a child route, e.g.:

```
/**: {
  kinde: {
    permissions: ['example_permission'],
  }
}
```

(this would apply to all routes but wouldn't be detected by this code.)

danielroe replied:
I would also recommend using a boolean object:

```ts
{
  'example_permission': true
}
```

as this is more easily merged - otherwise arrays just increase in size, rather than allow you to turn off a flag

DanielRivers replied:
@danielroe This has been resolved now.

DanielRivers commented:
@danielroe I added the ability to define public routes if a global rule is defined but would like to keep a single route as public,  I also added ability to define the redirect as external to simplify if wanting to call the login api route.

Ready for another review
